src/proxy/http2/Http2ConnectionState.cc

```diff
@@ -1434,6 +1451,14 @@ Http2ConnectionState::main_event_handler(int event, void *edata)
 
     ink_assert(this->fini_received == false);
     this->fini_received = true;
+
+    if (shutdown_state == HTTP2_SHUTDOWN_INITIATED) {
+      if (shutdown_cont_event) {
+        shutdown_cont_event->cancel();
+      }
+      this->_shutdown_cont();
+    }
```

Comment:
So cleaning up the in progress shutdown events, since we just shutdown cleanly here?

Reply:
Yes, since we already started a grace shutdown, I think we should finish the whole process.

Reply:
I still don't get why we need to call `finish_grace_shutdown()` here.
When we schdule `HTTP2_SESSION_EVENT_FINI`, we send `GOAWAY` with `latest_streamid_in`, right? Doesn't this send an extra `GOAWAY`?

- GOAWAY with the big number for graceful shutdown
- GOAWAY with latest_streamid_in at the place we schedule `HTTP2_SESSION_EVENT_FINI`
- GOAWAY with latest_streamid_in here in the `HTTP2_SESSION_EVENT_FINI` handler .

Reply:
Because when we get a `HTTP2_SESSION_EVENT_FINI` we set the handler to `Http2ConnectionState::state_closed`, which just does `shutdown_cont_event = nullptr`, so the GOAWAY with `latest_streamid_in` never gets sent out.

Reply:
We only want to send two GOAWAY frames at most, A) one with the big number, and B) one with `latest_streamid_in`, right?

B is sent when we schedule `HTTP2_SESSION_EVENT_FINI` event. For example:
https://github.com/apache/trafficserver/blob/c49e60f769db8bdd2b662a2fa1fd2f90108a47dd/src/proxy/http2/Http2ConnectionState.cc#L1451-L1455

So when we reach this line, which is in `HTTP2_SESSION_EVENT_FINI` handler, B is already sent (or it's in the send buffer at least). Calling `finish_grace_shutdown` would send third GOAWAY.

Reply:
yes, but that's just one case, but not for all of them. That's why I check for the `shutdown_state == HTTP2_SHUTDOWN_INITIATED` to make sure that we only send a second GOAWAY with `latest_streamid_in` when we initiated a grace shutdown.

Reply:
The only two places where we don't send GOAWAY prior to scheduling `HTTP2_SESSION_EVENT_FINI` are `Http2ClientSession::do_io_close` and `Http2ServerSession::do_io_close`. And those places seem to be too late to send frames anyway because `connection_state.is_state_closed()` is true if `HTTP2_SESSION_EVENT_FINI` is scheduled there.

Even if I miss a place that schedules `HTTP2_SESSION_EVENT_FINI` without sending GOAWAY, or even if we add new code like that, checking `shutdown_state == HTTP2_SHUTDOWN_INITIATED` doesn't make sense to me. How does it prevent sending a third `GOAWAY`? It's unrelated to whether a GOWAY with `latest_streamid_in` is sent. Are you saying the change ensures that there will be at least one GOAWAY frame with `latest_streamid_in` and there may be two of them in some cases, and that's fine?

I hate to add flags just to correct behavior forcibly, but it would be more straightforward solution if we add `bool is_the_last_goaway_sent` and set `true` in `Http2ConnectionState::send_goaway_frame` if the stream id is the big number. Then we can use it to ensure that ATS always send just one GOAWAY frame with `latest_streamid_in`.

Reply:
The only way we get an event `HTTP2_SESSION_EVENT_SHUTDOWN_INIT` is when `shutdown_state == HTTP2_SHUTDOWN_NOT_INITIATED`, and that is the trigger for a grace shutdown since we only schedule that event when checking for the state like this. 
```
if (this->connection_state.get_shutdown_state() == HTTP2_SHUTDOWN_NOT_INITIATED) {
  send_connection_event(&this->connection_state, HTTP2_SESSION_EVENT_SHUTDOWN_INIT, this);
}
```

And from testing I've done so far, I was never able to get the second GOAWAY frame as described in the RFC when a grace shutdown is initiated, and that's why I created this PR. But I can move `this->_finish_grace_shutdown();` two lines above where it is at now, since `shutdown_cont_event` is only scheduled during a grace shutdown.

Reply:
Maybe some other changes in the past few months have fixed this in a different way then, I just checked with Proxy Verifier and it is also getting the second GOAWAY frame now. So it's your call then, do you want to add a flag to make sure the second GOAWAY gets sent during a grace shutdown, or I can remove this part of the change and just add the test to check things are working as intended.

Reply:
Okay, nice. Then the current code works as I expect. I would not add the flag, because I think something is wrong if the flag has to work.

Let's just add the test. Thank you for writing the test.

Reply:
Sounds good.